<a href="https://colab.research.google.com/github/DavidSenseman/BIO1173/blob/main/Regression_Spring25_Assigment_01_BoneMarrow.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

---------------------------
**COPYRIGHT NOTICE:** This Jupyterlab Notebook is a Derivative work of [Jeff Heaton](https://github.com/jeffheaton) licensed under the Apache License, Version 2.0 (the "License"); You may not use this file except in compliance with the License. You may obtain a copy of the License at

> [http://www.apache.org/licenses/LICENSE-2.0](http://www.apache.org/licenses/LICENSE-2.0)

Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.

------------------------

# **BIO 1173: Intro Computational Biology**

**Assignment 1: Neural Networks for Analysis of Tabular Data**

* Instructor: [David Senseman](mailto:David.Senseman@utsa.edu), [Department of Integrative Biology](https://sciences.utsa.edu/integrative-biology/), [UTSA](https://www.utsa.edu/)



# **READ CAREFULLY**

The **_first_**  digit in your myUTSA ID (e.g. "abc123") will determine which dataset you are to analyze for this assignment and which type of neural network (i.e. classification or regression) you will need to construct. For example, if your myUTSA ID was **vue682**, then your first digit is the number `6`.

**---WARNING------WARNING------WARNING------WARNING------WARNING------WARNING---**

You are **not** free to choose any dataset for this assignment. If analyze the wrong dataset, your assignment will **NOT BE GRADED**. If you are uncertain which dataset you should be working on, contact your Instructor for help. Remember, your score in this assignment will have a large impact on your course grade so please be careful.


| First Digit myUTSA ID    | Dataset to Analyze      | Neural Network Type
--------------------------|-------------------------|-----------------
0                         | Hepatitis               | Binary Classification
1                         | Coimbra Breast Cancer   | Binary Classification
2                         | Parkinson Speech        | Binary Classification
3                         | Indian Liver            | Binary Classification
4                         | Thyroid Replacement     | Multiclass Classification
5                         | Wine Quality            | Multiclass Classification
6                         | Liver Disease           | Multiclass Classification
7                         | Bone Marrow Transplant  | Regression
8                         | German Breast Cancer    | Regression
9                         | Diabetes Progression    | Regression

#### **NOTE: You can only use this Colab notebook if the first digit of your _myUTSA_ ID  is between `7` and `9`.**

## **MAKE A COPY OF THIS NOTBOOK!**

For your assignment to be graded, you **must** make a copy of this Colab notebook in your GDrive and you **must** use this copy as your worksheet.

## Google CoLab Instructions

You MUST run the following code cell to get credit for this class lesson. By running this code cell, you will map your GDrive to /content/drive and print out your Google GMAIL address. Your Instructor will use your GMAIL address to verify the author of this class lesson.

In [ ]:
# YOU MUST RUN THIS CELL FIRST

try:
    from google.colab import drive
    drive.mount('/content/drive', force_remount=True)
    from google.colab import auth
    auth.authenticate_user()
    COLAB = True
    print("Note: using Google CoLab")
    import requests
    gcloud_token = !gcloud auth print-access-token
    gcloud_tokeninfo = requests.get('https://www.googleapis.com/oauth2/v3/tokeninfo?access_token=' + gcloud_token[0]).json()
    print(gcloud_tokeninfo['email'])
except:
    print("**WARNING**: Your GMAIL address was **not** printed in the output below.")
    print("**WARNING**: You will NOT receive credit for this assignment.")
    COLAB = False

Mounted at /content/drive
Note: using Google CoLab
david.senseman@gmail.com


# **The Purpose of Assignments**

In this course, **_Assignments_** are designed to help me (and you) assess your ability to transfer knowledge gained in completing class coding exercises to solving more realistic problems.

Assignments play a pivotal role in reinforcing your learning, as they require you to apply theoretical concepts to practical scenarios. This helps solidify your understanding and enhances your problem-solving skills. By tackling these assignments independently, you develop critical thinking and the ability to synthesize information from various sources. Moreover, assignments encourage you to explore topics more deeply, fostering intellectual curiosity and promoting a deeper engagement with the subject matter. Ultimately, these assignments are not just a measure of your learning, but a means to equip you with the skills needed for real-world applications and future challenges.

Your GMAIL address **must** appear in the output in order for your work to be graded.

# **Assigment 1: Regression**

**Assignment_01** is specifically designed to assess your ability to write the Python/Tensorflow/Keras code necessary to build neural networks that can analyze tabular data stored in a Pandas DataFrame. These analyzes include: (1) binary classification, (2) multiclass classification and (3) regression.

You will use this Colab notebook **only** if the first digit in your myUTSA ID is between `7` and `9`. If that is correct, you have been assigned to perform **regression**.

Unlike your class lessons, you will **not** be given examples that you can use to simply copy-and-paste code. Rather, you will be given a problem to solve and it will be up to you to use code snippets that you have been given previously to solve different aspects of this assignment. And unlike your class lessons, your will **not** be given the correct output. In other words, this assignment is basically how you would solve an actual biomedical problem.


### Define functions

The cell below creates several functions that are needed for this assignment. If you don't run this cell, you will receive errors later when you try to run some cells.

# **Regression by Neural Networks**
**Regression** of tabular data is a type of supervised learning task where the goal is to predict a continuous target variable based on the attributes of the observations.

#### **Regression:**
Regression deals with problems where the outcome is a continuous variable. Examples include:
1. **Medical Predictions:**
   - Predicting blood pressure levels based on patient attributes.
   - Estimating the progression of a disease over time.
   - Forecasting patient recovery times after surgery.
   - Estimating the cost of medical treatments.

2. **Biological Predictions:**
   - Predicting enzyme activity levels based on substrate concentration.
   - Estimating growth rates of bacteria or cell cultures under different conditions.
   - Forecasting gene expression levels based on environmental factors.
   - Estimating the concentration of hormones in blood samples.

3. **Disease Progression:**
   - Predicting tumor growth rates based on imaging data.
   - Estimating the severity of disease symptoms over time.
   - Forecasting the spread of infectious diseases in a population.
   - Estimating the decline in lung function in patients with respiratory diseases.

4. **Treatment Outcomes:**
   - Predicting patient response to different medication dosages.
   - Estimating the duration of hospital stays based on initial symptoms.
   - Forecasting the likelihood

# **Descriptions of Data Sets for Regression**

This section describes the various datasets, information for downloading them, and what variable(s) your network should predict. Remember, you do **not** earn and credit if you analyze the wrong dataset. Pay particular attention to the **variable for Regression** for your assigned dataset. You will need to know the name of this feature when you are constructing yor `X-` and `Y-feature` vectors.


-----------------------------


## **Bone Marrow Transplant - 1st myUTSA Digit = 7**

#### **Filename:** `bone_marrow_transplant.csv`
#### **Response Variable for Regression (Y):** `survival_time`

### **Bone Marrow Transplant dataset**

The **Bone Marrow Transplant** dataset describes pediatric patients with several hematologic diseases: malignant disorders (e.g. patients with acute lymphoblastic leukemia, with acute myelogenous leukemia, with chronic myelogenous leukemia, with myelodysplastic syndrome) and nonmalignant cases (i.a. patients with severe aplastic anemia, with Fanconi anemia, with X-linked adrenoleukodystrophy).

All patients were subject to the unmanipulated allogeneic unrelated donor hematopoietic stem cell transplantation.

The motivation of this study was to identify the most important factors influencing the success or failure of the transplantation procedure. In particular, verification of the research hypothesis that increased dosage of CD34+ cells / kg extends overall **survival time** without simultaneous occurrence of undesirable events affecting patients' quality of life.

#### Features:

1. **Recipientgender:** Male, Female
- **Stemcellsource:** Source of hematopoietic stem cells (Peripheral blood - 1	 Bone marrow - 0)
2. **Donorage:** Age of the donor at the time of hematopoietic stem cells apheresis
3. **Donorage35:** - Donor age <35 - 0	 Donor age >=35 - 1
- **IIIV:** - Development of acute graft versus host disease stage II or III or IV (Yes - 1	 No - 0)
4. **Gendermatch:** Compatibility of the donor and recipient according to their gender (Female to Male - 1	 Other - 0)
- **DonorABO:** ABO blood group of the donor of hematopoietic stem cells (0 - 0	1	 A	 B=-1	 AB=2)
5. **RecipientABO:** ABO blood group of the recipient of hematopoietic stem cells (0 - 0	1	 A	 B=-1	 AB=2)
6. **RecipientRh:** Presence of the Rh factor on recipient�s red blood cells ('+' - 1	 '-' - 0)
7. **ABOMatch:** Compatibility of the donor and the recipient of hematopoietic stem cells according to ABO blood group (matched - 1	 mismatched - 1)
8. **CMVstatus:** Serological compatibility of the donor and the recipient of hematopoietic stem cells according to cytomegalovirus
infection prior to transplantation (the higher the value the lower the compatibility)
9. **RecipientCMV:** Presence of cytomegalovirus infection in the donor of hematopoietic stem cells prior to transplantation (presence - 1	 absence - 0)
10. **Disease:** Type of disease (ALL	AML	chronic	nonmalignant lymphoma)
11. **Riskgroup:** High risk - 1	 Low risk - 0
12. **Txpostrelapse:** The second bone marrow transplantation after relapse (No - 0; Yes - 1)
13. **Diseasegroup:** Type of disease (malignant - 1	 nonmalignant - 0)
14. **HLAmatch:** Compatibility of antigens of the main histocompatibility complex of the donor and the recipient of hematopoietic stem cells					according to ALL international BFM SCT 2008 criteria
15. **HLAmismatch:** - HLA matched - 0	 HL mismatched - 1
16.**Antigen:** In how many anigens there is difference beetwen the donor nad the recipient (-1 - no differences	 0 - one difference	1 (2) - two (three) diffences)
17. **Allel:** In how many allele there is difference beetwen the donor nad the recipient {-1 no differences	0 - one difference	 1 (2) (3) - two	 (tree	 four) differences)
18. **HLAgrI:** The differecne type beetwien the donor and the recipient (HLA mateched - 0	the difference is in only one antigen - 1, the difference is only in one allel - 2, the difference is only in DRB1 cell - 3, two differences (two allele or two antignes) - 4, two differences (two allele or two antignes) - 5)
19. **Recipientage:** Age of the recipient of hematopoietic stem cells at the time of transplantation
20. **Recipientage10:**  Recipient age <10 - 0	 Recipient age>=10 - 1
21. **Recipientageint:** Recipient age in (0	5] - 0	 (5	 10] - 1	 (10	 20] - 2
22. **Relapse:** Reoccurrence of the disease (No - 0	 Yes - 1)
23. **aGvHDIIIIV:** Development of acute graft versus host disease stage III or IV (Yes - 0	 No - 1)
24. **extcGvHD:** Development of extensive chronic graft versus host disease (Yes - 0	 No - 1)
25. **CD34kgx10d6:** CD34+ cell dose per kg of recipient body weight (10^6/kg)
26. **CD3dCD34:** CD3+ cell to CD34+ cell ratio
27. **CD3dkgx10d8:** CD3+ cell dose per kg of recipient body weight (10^8/kg)
28. **Rbodymass:** Body mass of the recipient of hematopoietic stem cells at the time of transplantation
29. **ANCrecovery:** Time to neutrophils recovery defined as neutrophils count >0.5 x 10^9/L
30. **PLTrecovery:**  Time to platelet recovery defined as platelet count >50000/mm3
31. **time_to_aGvHD_III_IV:** Time to development of acute graft versus host disease stage III or IV
32. **survival_time:** Survival time in days -- Response Variable (Y values)
33. **survival_status:** Survived (Yes - 0 No - 1)


------------------------------------------

## **German Breast Cancer Dataset - 1st myUTSA Digit = 8**

#### **Filename:** `GermanBreastCancer.csv`
#### **Response Variable for Regression (Y):** `time`

### **German Breast Cancer Study Group (GBSG2) dataset**

The **German Breast Cancer Study Group (GBSG2)** dataset studies the effects of hormone treatment on recurrence-free survival time. The event of interest is the recurrence of cancer time. This data frame contains the observations of 686 women.

#### Features:

1.**horTh:** hormonal therapy, a factor at two levels (yes 1 and no 2).
2. **age:** age of the patients in years.
3. **menostat:** menopausal status, a factor at two levels pre (premenopausal = 1) and post (postmenopausal = 2).
4. **tsize:** tumor size (in mm).
5. **tgrade:** tumor grade, a ordered factor at levels 1 < 2 < 3.
6. **pnodes:** number of positive nodes.
7. **progrec:** progesterone receptor (in fmol).
8. **estrec:** estrogen receptor (in fmol).
9. **time:** recurrence free survival time (in days). Response variable (Y values)
10. **cens:** censoring indicator (0- censored, 1- event).


-----------------------------------------

## **Diabetes Progression Dataset - 1st myUTSA Digit = 9**

#### **Filename:** `diabetes_progrssion.csv`

#### **Response Variable for Regression (Y):** `disease_progression`

**Diabetes** is more than just a high blood sugar condition; it’s a complex, chronic illness that, if not managed properly, can lead to severe health complications like heart disease, nerve damage, and kidney failure. Recognizing the urgency of effective management, I embarked on a project to predict diabetes progression using advanced data science techniques. This project involved regression analysis, data visualization, and model evaluation. The complete code is available on my GitHub page, aiding better medical decision-making and patient care.

**Dataset Overview**

The Diabetes dataset, a mainstay in regression analysis, includes ten baseline variables such as age, sex, BMI, average blood pressure, and six blood serum measurements for 442 diabetes patients. The target variable is a quantitative measure of disease progression one year after the baseline.

**Data Characteristics:**
**Number of Instances:** 442
**Number of Attributes:** 10 numeric predictive values
**Target:** Quantitative measure of disease progression

**Attribute Information:**
- **age:** Age in years
- **sex:** Gender of the patient
- **bmi:** Body mass index
- **bp:** Average blood pressure
- **serum_chol:** Total serum cholesterol (mg/dL)
- **ldl:** Low-density lipoproteins (mg/dL)
- **hdl:** High-density lipoproteins (mg/dL)
- **chol_hdl_ratio:** Total cholesterol / HDL
- **log_trigly:** Log of serum triglycerides level
- **blood_glu:** Blood sugar level (mg/dL)
- **disease_progression:** Disease progression one year after baseline. Response variable (Y values)

-----------------------------------

# **General Instructions**

To make the assignment mpre manageable, you will given a number of specific steps to perform. Generally, in each step you are given a class lesson that you can use for a reference. For example, in **Step 1: Download and Extract Data** you are given **REF: Class_01_6**. That means Class_01_6 has an example of the code that you could use to complete that step.

**Variable Names**

In writing your code for this assignment, you are free to give your variables any name that makes sense to you, this includes the name of the DataFrame that holds your data. When it has been necessary to give an example name for a DataFrame in an instruction, the DataFrame has been called `dataFrameDF`. If you copy-and-paste instructions that use `dataFrameDF` into your code cell, you will need to edit the name to match the name you have given to your DataFrame.

### **Can I Use AI?**

You are free to use AI (e.g. Microsoft Co-Pilot) to help you complete your assignment---but you really need to be careful.

While AI can be very helpful in correcting coding errors, it can also give you code that is totally incorrect for this assignment. If your aren't sure what you are doing, it's much, much safer to get help with any of your coding problems from your course instructor and/or course TA's.

In [1]:
# Create functions for this lesson

def list_float_columns(dataframe):
    """
    Create a list of all columns in a DataFrame that contain float values.

    Parameters:
    dataframe (pd.DataFrame): The DataFrame to check.

    Returns:
    list: A list of column names that contain float values.
    """
    float_columns = [col for col in dataframe.columns if dataframe[col].dtype == 'float64']
    return float_columns

## **STEP 1: Download and Extract Data**

**REF: Class_01_6 (Example 1)**

As usual, a coding project starts with downloading a dataset. Since your dataset is in tabular form, you should use Pandas to read the datafile and store the information in a DataFrame. You are free to choose the name for all or your variables in this assignment, including the name of your DataFrame.

In the cell below, write the code to download your datafile from the course server and create a Pandas DataFrame to store your data. Use the function `display()` to show the data in 8 rows. For full credit, you need to show **all** of the columns in your DataFrame. You can do this by using the following code:

~~~text
# Set max columns and max rows
pd.set_option('display.max_columns', dataFrameDF.shape[1])
pd.set_option('display.max_rows', 8)
~~~

Remember, you will need to edit the name `dataFrameDF` to match the name you select for your DataFrame.


In [2]:
# Step 1: Download and Extract Data

import numpy as np
import pandas as pd

# Read file and create DataFrame
bmDF = pd.read_csv(
    "https://biologicslab.co/BIO1173/data/bone_marrow_transplant.csv",
#    index_col=0,
    na_values=['NA','?'])

# Set max columns and max rows
pd.set_option('display.max_columns', bmDF.shape[1])
pd.set_option('display.max_rows', 8)

# Display DataFrame
display(bmDF)

Recipientgender  Stemcellsource   Donorage  Donorage35  IIIV  Gendermatch  \
0              Male               1  22.830137           0     1            0   
1              Male               0  23.342466           0     1            0   
2              Male               0  26.394521           0     1            0   
3            Female               1  33.358904           0     0            0   
..              ...             ...        ...         ...   ...          ...   
138            Male               1  37.575342           1     1            0   
139            Male               1  22.895890           0     0            0   
140          Female               1  27.347945           0     1            0   
141          Female               1  27.780822           0     1            0   

     DonorABO  RecipientABO  RecipientRh  ABOmatch  CMVstatus  DonorCMV  \
0           1             1            1         0          3         1   
1          -1            -1            1         0          0         0   
2          -1            -1            1         0          2         0   
3           1             2            0         1          0         0   
..        ...           ...          ...       ...        ...       ...   
138         1             1            0         0          3         1   
139         1             0            1         1          2         0   
140         1            -1            1         1          3         1   
141         1             0            1         1          0         0   

     RecipientCMV  Riskgroup  Txpostrelapse  Diseasegroup  HLAmatch  \
0               1          1              0             1         0   
1               0          0              0             1         0   
2               1          0              0             1         0   
3               1          1              0             1         1   
..            ...        ...            ...           ...       ...   
138             1          1              1             1         0   
139             1          1              0             1         0   
140             1          0              0             0         0   
141             0          0              0             1         1   

     HLAmismatch  Antigen  Alel  HLAgrI  Recipientage  Recipientage10  \
0              0       -1    -1       0           9.6               0   
1              0       -1    -1       0           4.0               0   
2              0       -1    -1       0           6.6               0   
3              0        1     0       1           1.3               0   
..           ...      ...   ...     ...           ...             ...   
138            0       -1    -1       0          12.9               1   
139            0       -1    -1       0          13.9               1   
140            0       -1    -1       0          10.4               1   
141            0        1     0       1           8.0               0   

     Recipientageint  Relapse  aGvHDIIIIV  extcGvHD  CD34kgx10d6   CD3dCD34  \
0                  1        0           0         1         7.20   1.338760   
1                  0        1           1         1         4.50  11.078295   
2                  1        1           1         1         7.94  19.013230   
3                  0        0           1         1        51.85   3.972255   
..               ...      ...         ...       ...          ...        ...   
138                2        1           0         1        11.08   2.522750   
139                2        0           1         1         4.64   1.038858   
140                2        0           1         1         7.73   1.635559   
141                1        0           0         0        15.41   8.077770   

     CD3dkgx10d8  Rbodymass  ANCrecovery  PLTrecovery  time_to_aGvHD_III_IV  \
0           5.38       35.0           19           51                    32   
1           0.41       20.6           16           

If your code is correct you should see a table with a relatively large number of columns that may extend beyond the right edge of your notebook.

## **STEP 2: Describe DataFrame**

**REF: Class_01_6 (Example 3)**

The `df.describe()` command in Python is used with pandas DataFrames. It provides a summary of statistics for each column in the DataFrame. By default, it will return the count, mean, standard deviation, min, 25th percentile (Q1), median (50th percentile), 75th percentile (Q3), and max values for numerical columns. It can be a handy tool for getting a quick overview of your dataset!

Use the `df.describe()` command to summaries the data in your DataFrame. Make sure to replace the `df` with the actual name of your DataFrame.

Again use these commands to set the display options:

~~~text
# Set max columns and max rows
pd.set_option('display.max_columns', dataFrameDF.shape[1])
pd.set_option('display.max_rows', 8)
~~~


In [3]:
# Step 2: Describe DataFrame

import pandas as pd

# Set max columns and max rows
pd.set_option('display.max_columns', bmDF.shape[1])
pd.set_option('display.max_rows', 8)

# Describe() method
bmDF.describe()

Stemcellsource    Donorage  Donorage35        IIIV  Gendermatch  \
count      142.000000  142.000000  142.000000  142.000000   142.000000   
mean         0.816901   32.931082    0.408451    0.612676     0.169014   
std          0.388116    8.007697    0.493287    0.488863     0.376091   
min          0.000000   18.646575    0.000000    0.000000     0.000000   
25%          1.000000   26.461644    0.000000    0.000000     0.000000   
50%          1.000000   31.708219    0.000000    1.000000     0.000000   
75%          1.000000   40.267808    1.000000    1.000000     0.000000   
max          1.000000   52.476712    1.000000    1.000000     1.000000   

         DonorABO  RecipientABO  RecipientRh    ABOmatch   CMVstatus  \
count  142.000000    142.000000   142.000000  142.000000  142.000000   
mean     0.366197      0.260563     0.830986    0.704225    1.535211   
std      0.837752      0.957831     0.376091    0.458006    1.146453   
min     -1.000000     -1.000000     0.000000    0.000000    0.000000   
25%      0.000000     -1.000000     1.000000    0.000000    0.000000   
50%      0.000000      0.000000     1.000000    1.000000    2.000000   
75%      1.000000      1.000000     1.000000    1.000000    2.000000   
max      2.000000      2.000000     1.000000    1.000000    3.000000   

         DonorCMV  RecipientCMV   Riskgroup  Txpostrelapse  Diseasegroup  \
count  142.000000    142.000000  142.000000     142.000000    142.000000   
mean     0.401408      0.584507    0.366197       0.119718      0.823944   
std      0.491919      0.494551    0.483470       0.325781      0.382216   
min      0.000000      0.000000    0.000000       0.000000      0.000000   
25%      0.000000      0.000000    0.000000       0.000000      1.000000   
50%      0.000000      1.000000    0.000000       0.000000      1.000000   
75%      1.000000      1.000000    1.000000       0.000000      1.000000   
max      1.000000      1.000000    1.000000       1.000000      1.000000   

         HLAmatch  HLAmismatch     Antigen        Alel      HLAgrI  \
count  142.000000   142.000000  142.000000  142.000000  142.000000   
mean     0.676056     0.147887   -0.063380   -0.246479    1.232394   
std      0.776876     0.356245    0.997975    0.876701    1.691302   
min      0.000000     0.000000   -1.000000   -1.000000    0.000000   
25%      0.000000     0.000000   -1.000000   -1.000000    0.000000   
50%      1.000000     0.000000    0.000000    0.000000    1.000000   
75%      1.000000     0.000000    1.000000    0.000000    2.000000   
max      3.000000     1.000000    2.000000    3.000000    7.000000   

       Recipientage  Recipientage10  Recipientageint     Relapse  aGvHDIIIIV  \
count    142.000000      142.000000       142.000000  142.000000  142.000000   
mean       9.458451        0.450704         1.190141    0.176056    0.802817   
std        5.042982        0.499325         0.833119    0.382216    0.399280   
min        0.700000        0.000000         0.000000    0.000000    0.000000   
25%        4.925000        0.000000         0.000000    0.000000    1.000000   
50%        9.100000        0.000000         1.000000    0.000000    1.000000   
75%       13.475000        1.000000         2.000000    0.000000    1.000000   
max       20.200000        1.000000         2.000000    1.000000    1.000000   

         extcGvHD  CD34kgx10d6    CD3dCD34  CD3dkgx10d8   Rbodymass  \
count  142.000000   142.000000  142.000000   142.000000  137.000000   
mean     0.838028    12.609246    5.015907     4.841408   34.118248   
std      0.369729     9.900743    9.740032     3.693064   17.972121   
min      0.000000     0.790000    0.309048     0.040000    6.400000   
25%      1.000000     6.425000    1.831103     1.937500   19.000000   
50%      1.000000    10.070000    2.727613     4.465000   30.500000   
75%      1.000000    16.015000    5.134882     6.785000   49.000000   
max      1.000000    57.780000   99.560970    20.020000   76.000000   

       ANCreco

If your code is correct you should see a table with a relatively large number of columns that may extend beyond the right edge of your notebook and 8 rows countaining the summary statistics for each column.

## **STEP 3: Find Missing Values**

**REF: Class_01_6 (Example 4)**

In **Biostatistics**, finding and replacing missing values is crucial for several reasons:

1. **Preserving Data Integrity**: Missing values can distort the analysis, leading to biased results. By addressing missing values, you ensure that your conclusions are based on complete and accurate data.

2. **Statistical Validity**: Many statistical tests and models require complete data. Missing values can reduce the statistical power of these tests, making it difficult to detect real effects.

3. **Avoiding Data Loss**: Simply discarding rows or columns with missing values can lead to a significant loss of data, especially if the dataset is already small. Imputing missing values helps retain as much data as possible.

4. **Model Accuracy**: Machine learning models can be sensitive to missing data. Handling missing values appropriately can improve the performance and accuracy of predictive models.

5. **Consistency**: Different columns in a dataset may have varying levels of missing data. Addressing these inconsistencies helps in creating a more uniform dataset, which is easier to analyze and interpret.

The `df.isnull()` command in pandas is used to detect missing values in a DataFrame. It returns a DataFrame of the same shape as the original, but with Boolean values: `True` where the value is missing (`NaN`) and `False` where the value is not missing.

In the cell below, use the command `df.isnull()` to find and print out any missing values in your DataFrame.

To make sure you see all of the values, use this code to set your display output:

~~~text
# Set max columns and max rows
pd.set_option('display.max_columns', dataFrameDF.shape[1])
pd.set_option('display.max_rows', dataFrameDF.shape[0])

~~~

In [4]:
# Step 3: Find Missing Values

import pandas as pd

# Find the locations of missing data
missing_locations = bmDF.isnull().any()

# Set max columns and max rows
pd.set_option('display.max_columns', bmDF.shape[1])
pd.set_option('display.max_rows', bmDF.shape[0])

# Display the locations of missing data
print(missing_locations)

Recipientgender         False
Stemcellsource          False
Donorage                False
Donorage35              False
IIIV                    False
Gendermatch             False
DonorABO                False
RecipientABO            False
RecipientRh             False
ABOmatch                False
CMVstatus               False
DonorCMV                False
RecipientCMV            False
Riskgroup               False
Txpostrelapse           False
Diseasegroup            False
HLAmatch                False
HLAmismatch             False
Antigen                 False
Alel                    False
HLAgrI                  False
Recipientage            False
Recipientage10          False
Recipientageint         False
Relapse                 False
aGvHDIIIIV              False
extcGvHD                False
CD34kgx10d6             False
CD3dCD34                False
CD3dkgx10d8             False
Rbodymass                True
ANCrecovery             False
PLTrecovery             False
time_to_aG

## **STEP 4: Replace Missing Values**

**REF: Class_01_6 (Example 5)**

One common strategy for replacing missing values is to use the `median` of the column to replace the missing value. The `median` is used instead of the column `mean` since the `median` is a robust measure of central tendency and is less affected by outliers compared to the `mean`.

To do this:
1. Calculate the median of the column: Use the `median()` function to find the median value of the column.

2. Replace the missing values: Use the `fillna()` function to replace the missing values with the median.

If your DataFrame had one (or more) columns with missing values, you will need write the Python code to replace these missing values with the column's `median` value in the cell below.

After you have replaced the missing values, again use the `df.isnull()` command to print out the names of columns with missing values to make sure all of this missing values have been taken care of. (Just use the same code you wrote in **Step 3**.)




In [5]:
import pandas as pd

# Find the median of the column Insulin
Rbodymass_med = bmDF['Rbodymass'].median()

# Print out the median value
print(f"The median value = {Rbodymass_med} for Rbodymass.")
print(f"Replacing missing values with {Rbodymass_med}.")

# Use fillna method
bmDF['Rbodymass'] = bmDF['Rbodymass'].fillna(Rbodymass_med)

# Find the locations of missing data
print("\nLooking for missing values...")  # The \n means print a newline
missing_locations = bmDF.isnull().any()

# Display the locations of missing data
print(missing_locations)

The median value = 30.5 for Rbodymass.
Replacing missing values with 30.5.

Looking for missing values...
Recipientgender         False
Stemcellsource          False
Donorage                False
Donorage35              False
IIIV                    False
Gendermatch             False
DonorABO                False
RecipientABO            False
RecipientRh             False
ABOmatch                False
CMVstatus               False
DonorCMV                False
RecipientCMV            False
Riskgroup               False
Txpostrelapse           False
Diseasegroup            False
HLAmatch                False
HLAmismatch             False
Antigen                 False
Alel                    False
HLAgrI                  False
Recipientage            False
Recipientage10          False
Recipientageint         False
Relapse                 False
aGvHDIIIIV              False
extcGvHD                False
CD34kgx10d6             False
CD3dCD34                False
CD3dkgx10d8             

## **STEP 5: Display Non-numeric Categories**

**REF: Class_02_2 (Example 3 - Step 1)**

When building neural networks it is especially important to know which columns contain non-numeric data ("strings").

In the cell below, write the code to print out a list of columns in your DataFrame that contain non-numeric data.

In [6]:
# Step 5: Display non-numeric categories

import pandas as pd

# Select columns
non_numerical_columns = bmDF.select_dtypes(exclude='number').columns

# Print result
print(*non_numerical_columns)

Recipientgender


## **STEP 6: Print Names in a Categorical Column**

**REF: Class_02_2 (Example 3 - Step 3)**

In the cell below, write the code to print out a list of categorical values (strings) that are in the non-numeric column found in **Step 5**.

In [7]:
# Step 6: Print names in a categorical column

import pandas as pd

# Generate a list with only unique values
numCat = list(bmDF['Recipientgender'].unique())

# Print out the results
print(f'Number of categories: {len(numCat)}')
print(f'String names: {numCat}')

Number of categories: 2
String names: ['Male', 'Female']


## **STEP 7: Map Strings to Integer Values**

**REF: Class_02_2 (Example 3 - Step 2)**

In the cell below, write the code to map each string shown in the output above to a different integer value. To make sure your mapping worked as intended, use the `display(df)` function to display your updated DataFrame.

In [8]:
# Step 7: Map Strings to Integer Values

# Set max columns and max rows
pd.set_option('display.max_columns', bmDF.shape[1])
pd.set_option('display.max_rows', 8)


# Define the mapping dictionary
mapping = {'Male': 1, 'Female': 2}

# Check if all values to be mapped are present in the column
unique_values = bmDF['Recipientgender'].unique()

# Find values in the column that are not in the mapping dictionary
missing_values = [value for value in unique_values if value not in mapping]

if missing_values:
    print(f"Error: The following values in the 'sex' column are not in the mapping dictionary: {missing_values}")
    print(f"Error: Either your mapping is wrong or you have already converted the strings to integers")
else:
    # Map the 'Gender' column using the mapping dictionary
    bmDF['Recipientgender'] = bmDF['Recipientgender'].map(mapping)
    print("Bone marrow transplant data after mapping:")
    display(bmDF)

Diabetes progression data after mapping:


Recipientgender  Stemcellsource   Donorage  Donorage35  IIIV  \
0                  1               1  22.830137           0     1   
1                  1               0  23.342466           0     1   
2                  1               0  26.394521           0     1   
3                  2               1  33.358904           0     0   
..               ...             ...        ...         ...   ...   
138                1               1  37.575342           1     1   
139                1               1  22.895890           0     0   
140                2               1  27.347945           0     1   
141                2               1  27.780822           0     1   

     Gendermatch  DonorABO  RecipientABO  RecipientRh  ABOmatch  CMVstatus  \
0              0         1             1            1         0          3   
1              0        -1            -1            1         0          0   
2              0        -1            -1            1         0          2   
3              0         1             2            0         1          0   
..           ...       ...           ...          ...       ...        ...   
138            0         1             1            0         0          3   
139            0         1             0            1         1          2   
140            0         1            -1            1         1          3   
141            0         1             0            1         1          0   

     DonorCMV  RecipientCMV  Riskgroup  Txpostrelapse  Diseasegroup  HLAmatch  \
0           1             1          1              0             1         0   
1           0             0          0              0             1         0   
2           0             1          0              0             1         0   
3           0             1          1              0             1         1   
..        ...           ...        ...            ...           ...       ...   
138         1             1          1              1             1         0   
139         0             1          1              0             1         0   
140         1             1          0              0             0         0   
141         0             0          0              0             1         1   

     HLAmismatch  Antigen  Alel  HLAgrI  Recipientage  Recipientage10  \
0              0       -1    -1       0           9.6               0   
1              0       -1    -1       0           4.0               0   
2              0       -1    -1       0           6.6               0   
3              0        1     0       1           1.3               0   
..           ...      ...   ...     ...           ...             ...   
138            0       -1    -1       0          12.9               1   
139            0       -1    -1       0          13.9               1   
140            0       -1    -1       0          10.4               1   
141            0        1     0       1           8.0               0   

     Recipientageint  Relapse  aGvHDIIIIV  extcGvHD  CD34kgx10d6   CD3dCD34  \
0                  1        0           0         1         7.20   1.338760   
1                  0        1           1         1         4.50  11.078295   
2                  1        1           1         1         7.94  19.013230   
3                  0        0           1         1        51.85   3.972255   
..               ...      ...         ...       ...          ...        ...   
138                2        1           0         1        11.08   2.522750   
139                2        0           1         1         4.64   1.038858   
140                2        0           1         1         7.73   1.635559   
141                1        0           0         0        15.41   8.077770   

     CD3dkgx10d8  Rbodymass  ANCrecovery  PLTrecovery  time_to_aGvHD_III_IV  \
0           5.38       35.0           19           51                    32   
1           0.41       20.6           16  

If your code is correct your output should be the same as Step 1 except that the strings in the non-numeric column should now be integers.

## **STEP 8: Shuffle and Reindex your DataFrame**

**REF: Class_02.3 (Example 2)**

**Shuffling and reindexing** the data are important steps when building neural networks for several reasons:

1. **Preventing Overfitting**: When the data is in a specific order, the model might learn patterns that are a result of the order rather than the actual data. Shuffling helps to prevent the model from overfitting to these spurious patterns.

2. **Improving Generalization**: By shuffling the data, you ensure that the model is exposed to a wide variety of examples during training, which helps it to generalize better to new, unseen data.

3. **Reducing Bias**: Shuffling ensures that the distribution of data is more uniform across the training batches. This reduces the risk of introducing bias, especially if the data has some inherent order that might otherwise affect the model's performance.

4. **Enhancing Convergence**: Neural networks often train faster and more reliably when the data is shuffled. This is because the model updates its parameters in a more representative and balanced manner.

5. **Ensuring Robustness**: By reindexing the data, you avoid potential issues that could arise from relying on the original indices, which might have some underlying structure or grouping that could affect the training process.


In the cell below, write the code to shuffle and reindex your DataFrame. Then display your shuffled DataFrame using the following display settings:

~~~text
# Set max columns and max rows
pd.set_option('display.max_columns', dataFrameDF.shape[1])
pd.set_option('display.max_rows', 8)
~~~

In [9]:
# Step 8: Shuffle and Reindex your DataFrame

import pandas as pd
import numpy as np

# Set the random seed to 42
np.random.seed(42)

# Use random.permutation function for shuffling & reindexing
bmDF = bmDF.reindex(np.random.permutation(bmDF.index))

# Set max columns and max rows
pd.set_option('display.max_columns', bmDF.shape[1])
pd.set_option('display.max_rows', 8)

# Display the shuffled & reindexed DataFrame
display(bmDF)

Recipientgender  Stemcellsource   Donorage  Donorage35  IIIV  \
131                2               1  41.476712           1     0   
19                 1               1  20.605479           0     0   
81                 1               1  35.764384           1     1   
96                 1               0  37.380822           1     1   
..               ...             ...        ...         ...   ...   
106                2               1  46.545205           1     0   
14                 1               1  43.578082           1     0   
92                 2               0  24.284932           0     1   
102                2               1  34.005479           0     1   

     Gendermatch  DonorABO  RecipientABO  RecipientRh  ABOmatch  CMVstatus  \
131            1        -1             2            1         1          1   
19             1         2             0            1         1          0   
81             0         0             1            1         1          0   
96             0         1             0            1         1          0   
..           ...       ...           ...          ...       ...        ...   
106            1         0            -1            0         1          0   
14             1         0             1            1         1          3   
92             0         0            -1            1         1          0   
102            0         0             0            1         0          0   

     DonorCMV  RecipientCMV  Riskgroup  Txpostrelapse  Diseasegroup  HLAmatch  \
131         1             0          1              1             1         0   
19          0             0          0              0             0         0   
81          0             0          0              0             1         1   
96          0             0          1              1             1         2   
..        ...           ...        ...            ...           ...       ...   
106         0             0          0              0             1         1   
14          1             1          1              1             1         1   
92          0             0          0              0             1         0   
102         0             0          0              0             1         2   

     HLAmismatch  Antigen  Alel  HLAgrI  Recipientage  Recipientage10  \
131            0       -1    -1       0          17.8               1   
19             0       -1    -1       0           8.5               0   
81             0        1     0       1           7.7               0   
96             1        2     0       4          17.8               1   
..           ...      ...   ...     ...           ...             ...   
106            0        1     0       1           6.0               0   
14             0        1     0       1          14.7               1   
92             0       -1    -1       0           1.3               0   
102            1        1     1       4           9.0               0   

     Recipientageint  Relapse  aGvHDIIIIV  extcGvHD  CD34kgx10d6  CD3dCD34  \
131                2        1           1         1         6.87  1.214932   
19                 1        0           1         1        12.95  1.672532   
81                 1        0           1         0         9.38  1.848795   
96                 2        0           1         0        10.85  1.070314   
..               ...      ...         ...       ...          ...       ...   
106                1        0           1         1        17.61  2.818993   
14                 2        0           1         1         0.82  0.368720   
92                 0        0           1         1         4.37  4.489022   
102                1        0           1         1         8.34  2.703806   

     CD3dkgx10d8  Rbodymass  ANCrecovery  PLTrecovery  time_to_aGvHD_III_IV  \
131         5.65       50.8           21           25               1000000   
19          7.74       26.8           13           1

If your code is correct, the output generated above by **Step 6** should look almost identical to the output you got after running **Step 7** except that this index numbers shown in the leftmore column should be in a random order.

## **STEP 9: Normalize Data to the Z-values**

**REF: Class_02_2 (Example 2)**

When building neural networks, it is important to normalize the data using Z-score normalization for several reasons:

1. **Standardization**: Z-score normalization standardizes the features to have a mean of 0 and a standard deviation of 1. This ensures that all features contribute equally to the model and prevents features with larger scales from dominating the learning process.

2. **Improved Convergence**: Neural networks often use gradient-based optimization algorithms. Normalized data can lead to more stable and faster convergence of these algorithms because the gradients are more balanced and less likely to explode or vanish.

3. **Enhanced Performance**: Normalized data often leads to better model performance. By bringing all features to a similar scale, the model can learn more effectively and produce more accurate predictions.

4. **Handling Outliers**: Z-score normalization reduces the impact of outliers by transforming the data to a standard scale. Outliers will have high positive or negative Z-scores, but their influence will be mitigated compared to unnormalized data.

5. **Consistency**: Normalization ensures that different features are on the same scale, which is particularly important when combining multiple datasets or using features with different units of measurement.

To make things easier, only normalize the `float` data in your dataset by converting these numbers into their Z-values.

In the cell below, use the function `list_float_columns()` to find the names of all of the columns in your DataFrame that contain float number. (NOTE: This is custom function that was created near the beginning of the assignment). Once you know the column names, use can use the `zscore()` function to normalize the values only in the columns containing floats. Print out the names of the columns containing floats, and the first five Z-scores in each of these columns.

In [10]:
# Step 9: Normalize Data to the Z-values

import pandas as pd
from scipy.stats import zscore

float_columns = list_float_columns(bmDF)
print(f"Columns with float values: {float_columns}")

for col in float_columns:
    bmDF[col] = zscore(bmDF[col])

# Print the first 5 values of each float column
for col in float_columns:
    print(f"First 5 values in column '{col}': {bmDF[col].head().tolist()}")

Columns with float values: ['Donorage', 'Recipientage', 'CD34kgx10d6', 'CD3dCD34', 'CD3dkgx10d8', 'Rbodymass']
First 5 values in column 'Donorage': [1.0709546066269835, -1.544668127283643, 0.35507476472429284, 0.55764985843136, 0.29533218217774526]
First 5 values in column 'Recipientage': [1.6599458990610154, -0.19072911512216453, -0.34992696580458854, 1.6599458990610154, 0.5455609442840467]
First 5 values in column 'CD34kgx10d6': [-0.581730342124011, 0.03453879584275109, -0.3273166026015482, -0.17831732088918975, -0.2756229742523626]
First 5 values in column 'CD3dCD34': [-0.3916239800114526, -0.34447630869759066, -0.3263154914143842, -0.4065243338796084, -0.2151638259669696]
First 5 values in column 'CD3dkgx10d8': [0.2197237410969829, 0.7876527768291075, 0.06211664027179993, 1.4398200905884853, -0.39711784316709525]
First 5 values in column 'Rbodymass': [0.9550130462767339, -0.40854825152003577, -0.8744316949339321, 0.9834205733141665, -0.18128803522057413]


**HINT:** If you see this error:

~~~text
---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
<ipython-input-8-9b058d8fbfea> in <cell line: 0>()
      4 from scipy.stats import zscore
      5
----> 6 float_columns = list_float_columns(dataFrameDF)
      7 print(f"Columns with float values: {float_columns}")
      8

NameError: name 'list_float_columns' is not defined
~~~~

it means you didn't run the cell called `Define functions` at the beginning of this assignment.


## **STEP 10: Pre-process Data for Neural Network Training**

**REF: Class_04_3 (Example 3)**

In the cell below, write the code to preprocess the data in your DataFrame to make ready to feed into your neural network.

NOTES:

1. Since you have already converted your float values into their Z-scores, do **not** normalize any data during your pre-processing. In other words, converting Z-scores into Z-scores, a second time, is not a good thing.

2. Basically all you need to do is write the code to generate your `X-feature vector` and your `Y-feature vector`. Your `Y-values` will be in the column that is the **response variable**. The name of the response variable (`Y` values) for your particular dataset was specified in the dataset description at the start of this assignment.

3. When generating your `X-feature vector`, you should use _all_ of the columns in your DataFrame **EXCEPT** for the column containing the `Y-values`.

4. Since you will be building a **Regression** neural network, **do not** one-hot encode the Y-values when generating your `Y-feature vector`. Instead, just use the values in column containing the Y-values.

5. Split data into training and test sets. You can use the code snippet in Example 3 in the following section:

~~~text
# Split data
~~~

6. When you are done, generating both your `X-` and `Y-` feature vectors, print out the first 4 values in each vector.

In [13]:
# Step 10 - Preprocess Data for Neural Network Training

import pandas as pd
import numpy as np
from scipy.stats import zscore
from sklearn.model_selection import train_test_split

# Generate column list for preprocessing
bmX_columns = bmDF.columns.drop('survival_time')

# Generate X feature vector
bmX = bmDF[bmX_columns].values
bmX = np.asarray(bmX).astype(np.float32)

# Generate Y feature vector
bmY = bmDF['survival_time'].values
bmY = np.asarray(bmY).astype(np.float32)

# Split data
bmX_train, bmX_test, bmY_train, bmY_test = train_test_split(
    bmX, bmY, test_size=0.25, random_state=42)

# Print out X and Y
np.set_printoptions(suppress=True,precision=4)
print("The first 4 X-values are:")
print(bmX[0:4])
print("\nTheir corresponding Y-values are:")
print(bmY[0:4])

The first 4 X-values are:
[[      2.           1.           1.071        1.           0.
        1.          -1.           2.           1.           1.
        1.           1.           0.           1.           1.
        1.           0.           0.          -1.          -1.
        0.           1.6599       1.           2.           1.
        1.           1.          -0.5817      -0.3916       0.2197
        0.955       21.          25.     1000000.           1.    ]
 [      1.           1.          -1.5447       0.           0.
        1.           2.           0.           1.           1.
        0.           0.           0.           0.           0.
        0.           0.           0.          -1.          -1.
        0.          -0.1907       0.           1.           0.
        1.           1.           0.0345      -0.3445       0.7877
       -0.4085      13.          11.     1000000.           0.    ]
 [      1.           1.           0.3551       1.           1.
        0. 

## **STEP 11: Construct and Compile Neural Network**

**REF: Class_04_3 (Example 4)**

In the cell below, use the Keras/Tensorflow libraries to construct and compile a regression neural network with 3 hidden layers but do **not** start your training in this step.

To construt your model, you can use the code chunk in Example 4 in the section that starts with:

~~~text
# Construct model-----------------------------------------------------
~~~
Since this neural network will perform regression, there should be only `1` neuron in the output layer.

After you construct it,  complie your model. To compile your model you can use the code chunk in Example 4 in the section that starts with:

~~~text
# Compile model------------------------------------------------------------------
~~~

Do **not** start training your model yet. This will be done in a separate step.

In [14]:
# Step 11: Construct and Compile Neural Network

from keras.models import Sequential
from keras.layers import Dense, Input
from keras.callbacks import EarlyStopping
from keras.optimizers import Adam
from keras.metrics import Precision, Recall

import numpy as np
from sklearn.model_selection import train_test_split

# Construct model---------------------------------------------------------------
bmModel = Sequential()
bmModel.add(Input(shape=(bmX.shape[1],)))
bmModel.add(Dense(100, activation='relu', kernel_initializer='random_normal'))
bmModel.add(Dense(50,activation='relu',kernel_initializer='random_normal'))
bmModel.add(Dense(25,activation='relu',kernel_initializer='random_normal'))
bmModel.add(Dense(1,activation='sigmoid',kernel_initializer='random_normal'))

# Compile model------------------------------------------------------------------
bmModel.compile(loss='mean_squared_error', optimizer='adam')



If your code is correct, you should **not** see any output after running the previous cell.

## **STEP 12: Print Summary of Your Model**

**REF: Class_04_2**

The `model.summary()` command in deep learning frameworks like Keras and TensorFlow provides a detailed summary of the neural network model. This summary includes useful information about the model's architecture, including:

1. **Layer Names and Types:** The name and type (e.g., Dense, Conv2D, LSTM) of each layer in the model.

2. **Output Shape:** The shape of the output produced by each layer.

3. **Number of Parameters:** The total number of trainable and non-trainable parameters in each layer. This includes both the weights and biases.

4. **Model Parameters Summary:** A total count of all trainable and non-trainable parameters in the model.

In the cell below, use the `model.summary()` command to print out the information about your neural network.



In [15]:
# Step 12: Print Summary of Your Model

bmModel.summary()

Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                         ┃ Output Shape                ┃         Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ dense (Dense)                        │ (None, 100)                 │           3,600 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_1 (Dense)                      │ (None, 50)                  │           5,050 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_2 (Dense)                      │ (None, 25)                  │           1,275 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_3 (Dense)                      │ (None, 1)                   │              26 │
└──────────────────────────────────────┴─────────────────────────────┴─────────────────┘

 Total params: 9,951 (38.87 KB)

 Trainable params: 9,951 (38.87 KB)

 Non-trainable params: 0 (0.00 B)

## **STEP 13: Create Early Stopping Monitor**

**REF: Class_04_2 (Example 9)**

An **Early Stopping Monitor** is a technique used during the training of neural networks to prevent overfitting and improve the model's generalization to new, unseen data. It works by monitoring the performance of the model on a validation dataset and stopping the training process when the performance starts to degrade.

#### Here’s how it works:

1. **Monitoring Performance**: Early stopping keeps track of a specific metric, such as validation loss or validation accuracy, during each epoch of training.

2. **Patience**: It has a parameter called "patience," which defines the number of epochs to wait for an improvement in the monitored metric before stopping the training. If the performance does not improve for a specified number of epochs, the training is stopped.

3. **Restore Best Weights**: In some implementations, early stopping can also restore the model weights to the state that resulted in the best performance on the validation set.

#### Benefits of early stopping include:

- **Preventing Overfitting**: By stopping training when the model starts to overfit the training data, early stopping helps maintain good generalization performance.
- **Saving Time and Resources**: It avoids unnecessary training epochs, saving computational resources and time.

In the cell below, write the code to create an Early Stopping Monitor that monitors `val_loss`. Set the parameter `patience` to `10`.

Example 9 in Class_04_2 show the code to create an Early Stopping Monitor. Do **not** copy all of the code in this example, just the code snippet in the section called:

~~~text
# Create EarlyStopping monitor--------------------------------------------
~~~



In [16]:
# Step 11: Create early stopping monitor

PATIENCE=10

bmMonitor = EarlyStopping(monitor='val_loss', min_delta=1e-3,
                        patience=PATIENCE, verbose=1, mode='auto',
                        restore_best_weights=True)

If your code is correct, you should not see any output.

## **Step 14: Train the Model**

**REF: Class_04_3 (Example 4)**

In the cell below, write the Python code to train the neural network that you constructed in **Step 9**. Set the number of epochs to `100`. Make sure the parameter `verbose` is set to `2` so that the output of each epoch is written out.

Example 4 in Class_04_3 shows the code to training your model. Do **not** copy all of the code in this example, just the code snippet in the section called:

~~~text
# Train model-------------------------------------------
~~~


In [17]:
# Step 14: Train the Model

# Set variables
EPOCHS=100
VERBOSE=2

# Train model
bmModel.fit(bmX_train,bmY_train,validation_data=(bmX_test,bmY_test),
          callbacks=[bmMonitor],verbose=VERBOSE,epochs=EPOCHS)



Epoch 1/100
4/4 - 2s - 589ms/step - loss: 1985967.3750 - val_loss: 1896304.6250
Epoch 2/100
4/4 - 0s - 26ms/step - loss: 1984764.1250 - val_loss: 1896296.6250
Epoch 3/100
4/4 - 0s - 36ms/step - loss: 1984759.8750 - val_loss: 1896287.5000
Epoch 4/100
4/4 - 0s - 32ms/step - loss: 1984755.0000 - val_loss: 1896277.5000
Epoch 5/100
4/4 - 0s - 35ms/step - loss: 1984749.3750 - val_loss: 1896266.8750
Epoch 6/100
4/4 - 0s - 35ms/step - loss: 1984742.3750 - val_loss: 1896253.3750
Epoch 7/100
4/4 - 0s - 37ms/step - loss: 1984734.7500 - val_loss: 1896235.1250
Epoch 8/100
4/4 - 0s - 35ms/step - loss: 1984723.0000 - val_loss: 1896213.1250
Epoch 9/100
4/4 - 0s - 25ms/step - loss: 1984709.7500 - val_loss: 1896184.8750
Epoch 10/100
4/4 - 0s - 37ms/step - loss: 1984694.5000 - val_loss: 1896150.8750
Epoch 11/100
4/4 - 0s - 33ms/step - loss: 1984673.2500 - val_loss: 1896114.6250
Epoch 12/100
4/4 - 0s - 37ms/step - loss: 1984657.3750 - val_loss: 1896076.5000
Epoch 13/100
4/4 - 0s - 33ms/step - loss: 198463

## **STEP 15: Compute MSE**

**REF: Class_04_3 (Example 5)**

In the cell below, write the code to compute the Mean Squared Error (MSE) for your model.

In [18]:
# Step 15: Compute MSE

import sklearn
from sklearn import metrics
from sklearn.metrics import mean_squared_error

# Use model to predict values
bmPred = bmModel.predict(bmX_test)

# Compare predicted and actual values
score = metrics.mean_squared_error(bmPred,bmY_test)

# Print results
print("Final score (MSE): {}".format(score))

2/2 ━━━━━━━━━━━━━━━━━━━━ 0s 62ms/step
Final score (MSE): 1895985.375


## **STEP 16: Compute RMSE**

**REF: Class_04_3 (Example 6)**

In the cell below, compute the Root Mean Squared Error (RMSE) for your neural network model. Print out the first 6 values in your `Y_compare` variable as well as your percent accuracy score.

NOTE: **Step 16** uses the variable holding the actual prediction values generated in **Step 15**, so you need to run **Step 15** before your run this step.

In [19]:
# Step 16: Compute RMSE

import numpy as np

# Compute RMSE
score = np.sqrt(metrics.mean_squared_error(bmPred,bmY_test))

# Print RSME
print("Final score (RMSE): {}".format(score))

# Print mean of Y
print(f"Average survival time:{bmY_test.mean()}")

# Print comparison
print(f"RMSE as percent of mean `survival_time`:  {score/bmY_test.mean()}")

Final score (RMSE): 1376.9478475962696
Average survival time:1149.27783203125
RMSE as percent of mean `survival_time`:  1.1980983268098302


## **Assignment Turn-in**

When you have completed and run all of the code cells, use the **File --> Print.. --> Save to PDF** to generate a PDF of your Colab notebook. Save your PDF as `Copy of Assignment_01.lastname.pdf` where _lastname_ is your last name, and upload the file to Canvas.

## **Poly-A Tail**

## **DeepSeek**

![__](https://upload.wikimedia.org/wikipedia/commons/thumb/e/ec/DeepSeek_logo.svg/1920px-DeepSeek_logo.svg.png)

**DeepSeek** (Chinese: 深度求索; pinyin: Shēndù Qiúsuǒ) is a Chinese artificial intelligence company that develops open-source large language models (LLMs). Based in Hangzhou, Zhejiang, it is owned and funded by Chinese hedge fund High-Flyer, whose co-founder, Liang Wenfeng, established the company in 2023 and serves as its CEO.

The DeepSeek-R1 model provides responses comparable to other contemporary large language models, such as OpenAI's GPT-4o and o1. It is trained at a significantly lower cost—stated at US \$6 million compared to \$100 million for OpenAI's GPT-4 in 2023—and approximately a tenth of the computing power used for Meta's comparable model, LLaMA 3.1. DeepSeek's AI models were developed amid United States sanctions on China and other countries for chips used to develop artificial intelligence, which were intended to restrict the ability of these countries to develop advanced AI systems. Lesser restrictions were later announced that would affect all but a few countries.

On 10 January 2025, DeepSeek released its first free chatbot app, based on the DeepSeek-R1 model, for iOS and Android; by 27 January, DeepSeek had surpassed ChatGPT as the most-downloaded free app on the iOS App Store in the United States, causing Nvidia's share price to drop by 18%. DeepSeek's success against larger and more established rivals has been described as "upending AI" and ushering in "a new era of AI brinkmanship". DeepSeek's compliance with Chinese government censorship policies and its data collection practices have also raised concerns over privacy and information control in the model, prompting regulatory scrutiny in multiple countries.

DeepSeek makes its generative artificial intelligence algorithms, models, and training details open-source, allowing its code to be freely available for use, modification, viewing, and designing documents for building purposes.However, reports indicate that the API version hosted in China applies content restrictions in accordance with local regulations, limiting responses on topics such as the Tiananmen Square massacre and Taiwan’s status. The company reportedly vigorously recruits young AI researchers from top Chinese universities, and hires from outside the computer science field to diversify its models' knowledge and abilities.

**Background**

In February 2016, High-Flyer was co-founded by AI enthusiast Liang Wenfeng, who had been trading since the 2007–2008 financial crisis while attending Zhejiang University. They began stock-trading with a deep learning model running on GPU on October 21, 2016. Prior to this, they used CPU-based models, mainly linear models. Most trading was done by AI by the end of 2017.

By 2019, he established High-Flyer as a hedge fund focused on developing and using AI trading algorithms. By 2021, High-Flyer exclusively used AI in trading, often using Nvidia chips. DeepSeek has made its generative artificial intelligence chatbot open source, meaning its code is freely available for use, modification, and viewing. This includes permission to access and use the source code, as well as design documents, for building purposes.

In 2021, while running High-Flyer, Liang began stockpiling Nvidia GPUs for an AI project.[20] According to 36Kr, Liang had built up a store of 10,000 Nvidia A100 GPUs, which are used to train AI, before the United States federal government imposed AI chip restrictions on China.

On 14 April 2023,[22] High-Flyer announced the start of an artificial general intelligence lab dedicated to research developing AI tools separate from High-Flyer's financial business. Incorporated on 17 July 2023, with High-Flyer as the investor and backer, the lab became its own company, DeepSeek. Venture capital firms were reluctant to provide funding, as they considered it unlikely that the venture would be able to generate an "exit" in a short period of time.

On May 16, 2023, the company Beijing DeepSeek Artificial Intelligence Basic Technology Research Co., Ltd. incorporated under the control of Hangzhou DeepSeek Artificial Intelligence Basic Technology Research Co., Ltd. As of May 2024, Liang Wenfeng held 84% of DeepSeek through two shell corporations.

After releasing DeepSeek-V2 in May 2024, which offered strong performance for a low price, DeepSeek became known as the catalyst for China's AI model price war. It was quickly dubbed the "Pinduoduo of AI", and other major tech giants such as ByteDance, Tencent, Baidu, and Alibaba began to cut the price of their AI models to compete with the company. Despite the low price charged by DeepSeek, it was profitable compared to its rivals that were losing money.

DeepSeek is focused on research and has no detailed plans for commercialization, which also allows its technology to avoid the most stringent provisions of China's AI regulations, such as requiring consumer-facing technology to comply with the government's controls on information.

DeepSeek's hiring preferences target technical abilities rather than work experience, resulting in most new hires being either recent university graduates or developers whose AI careers are less established. Likewise, the company recruits individuals without any computer science background to help its technology understand other topics and knowledge areas, including being able to generate poetry and perform well on the notoriously difficult Chinese college admissions exams (Gaokao).

**Training framework**

High-Flyer/DeepSeek has built at least two computing clusters, Fire-Flyer (萤火一号) and Fire-Flyer 2 (萤火二号). Fire-Flyer began construction in 2019 and finished in 2020, at a cost of 200 million yuan. It contained 1,100 GPUs interconnected at a rate of 200 Gbps. It was 'retired' after 1.5 years in operation. Fire-Flyer 2 began construction in 2021 with a budget of 1 billion yuan.[18] It was reported that in 2022, Fire-Flyer 2's capacity had been utilized at over 96%, totaling 56.74 million GPU hours. Of those GPU hours, 27% was used to support scientific computing outside the company.

Fire-Flyer 2 consisted of co-designed software and hardware architecture. On the hardware side, there are more GPUs with 200 Gbps interconnects. The cluster is divided into two "zones", and the platform supports cross-zone tasks. The network topology was two fat trees, chosen for its high bisection bandwidth. On the software side, there are

* **3FS (Fire-Flyer File System):** A distributed parallel file system. It was specifically designed for asynchronous random reads from a dataset, and uses Direct I/O and RDMA Read. In contrast to standard Buffered I/O, Direct I/O does not cache data. Caching is useless for this case, since each data read is random, and would not be reused.
* **hfreduce:** Library for asynchronous communication, originally designed to replace Nvidia Collective Communication Library (NCCL).[30] It was mainly used for allreduce, especially of gradients during backpropagation. It is asynchronously run on the CPU to avoid blocking kernels on the GPU.[28] It uses two-tree broadcast like NCCL.
* **hfai.nn:** Software library of commonly used operators in neural network training, similar to torch.nn in PyTorch.
* **HaiScale Distributed Data Parallel (DDP):** Parallel training library that implements various forms of parallelism in deep learning such as Data Parallelism (DP), Pipeline Parallelism (PP), Tensor Parallelism (TP), Experts Parallelism (EP), Fully Sharded Data Parallel (FSDP) and Zero Redundancy Optimizer (ZeRO). It is similar to PyTorch DDP, which uses NCCL on the backend.
* **HAI Platform:** Various applications such as task scheduling, fault handling, and disaster recovery.
During 2022, Fire-Flyer 2 had 5000 PCIe A100 GPUs in 625 nodes, each containing 8 GPUs. At the time, they chose to exclusively use PCIe instead of DGX version of A100, since at the time the models they trained could fit within a single 40 GB GPU VRAM, so there was no need for the higher bandwidth of DGX (i.e. they required only data parallelism but not model parallelism).[30] Later, they also incorporated NVLinks and NCCL, to train larger models that required model parallelism.